### PR TITLE
feat:  add POST /connect/sessions endpoint

### PR DESCRIPTION
**Pull request by TBonnin**
## Describe your changes

Adding the `POST /connect/sessions` endpoint as well as the `end_users` and `connect_sessions` tables.
Happy to discuss the naming 😺 

Spec:
```
POST /connect/sessions
-H "Authorization: Bearer PRIVATE_KEY"
{
  "end_user": {
     "id": "1jkhfz2FEuihojhfsdkjh29842", // Required
     "email": "a@a.com", // Required but for domain
     "display_name": "Mr A." // Optional
  },
   "organization": { // Optional
      "if": "foi", // Optional
      "display_name": "My org.", // Optional
   }
  "allowed_integrations": ['hubspot', 'github'] // Optional
  "integrations_config_defaults": { // Optional
    "hubspot": {
      "connection_config": {
        "extension": "fr",
      }
    },
  },
}

// Response
{
  "token": "nango_connect_session_dfkjdfkdjflksjfoizrjfozijfzoijfz",
  "expires_at": Date
}
```

`GET /connect/session` will come in next PR. Split them so PR are smaller

## Issue ticket number and link

https://linear.app/nango/issue/NAN-1775/%5Bserver%5D-api-endpoints-to-getpost-sessions

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/keystore/lib/utils/encryption.ts
Comment (bodinsamuel) [inline packages/keystore/lib/utils/encryption.ts]: If both call are mandatory, the throw could live in the getEncryptionKey imo

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]:
```suggestion
                }).strict(),
```

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]:
```suggestion
                        }).strict()
```

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]:
```suggestion
                        }).strict()
```

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]: opinion: you could reduce the indenting by reversing the condition

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]: At this stage I would not bother allowing just sending the `profileId` because it advocate for never sending the full profile again leading to desync

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]: if you want to keep it
```suggestion
                logger.error(getLinkedProfile.error);
```

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]: as we said in the meeting, profileId should probably be `companyId + profileId` 
if the email change then it needs update, if the companyId is different that means it's a different profile (even if it's the same end user)

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [inline packages/server/lib/controllers/connect/postSessions.ts]: I know the format is frustrating rn but I wouldn't use this, the error codes are entirely different from the rest of the API

File: packages/server/lib/services/connectSession.service.ts
Comment (bodinsamuel) [inline packages/server/lib/services/connectSession.service.ts]: We mostly use `DB` (except DbTask) we need to pick one

File: packages/server/lib/services/connectSession.service.ts
Comment (bodinsamuel) [inline packages/server/lib/services/connectSession.service.ts]: should take envId at least

File: packages/server/lib/services/connectSession.service.ts
Comment (bodinsamuel) [inline packages/server/lib/services/connectSession.service.ts]: same comment

File: packages/server/lib/services/linkedProfile.service.ts
Comment (bodinsamuel) [inline packages/server/lib/services/linkedProfile.service.ts]: same comment

File: packages/types/lib/connect/api.ts
Comment (bodinsamuel) [inline packages/types/lib/connect/api.ts]: opinion: we never discussed the shape of response, usually I'm advocating to use `data: {} | {}[]` to unify answer and to avoid potential conflict with prop that could be named `error`

File: packages/types/lib/linkedProfile/index.ts
Comment (bodinsamuel) [inline packages/types/lib/linkedProfile/index.ts]: nb: readonly is nice when you don't manipulate but can be annoying as lib user.
```ts
const lp: Partial<LinkedProfile> = {
    accountId: 1
}

// Not allowed bc it's readonly
lp.email = 'example@dev';
```

Not saying it's an issue in that case but just fyi

File: packages/keystore/lib/db/migrations/20240925111857_rename_connect_session.ts
Comment (bodinsamuel) [inline packages/keystore/lib/db/migrations/20240925111857_rename_connect_session.ts:1]: This migration does not run on my local setup

**Review by bodinsamuel — COMMENTED** [review]
Looks good 🚀 📶 

I can't test because of the migration that don't run (no error just don't run)
Few comments + others stuff (I guess you want to do all of that later but just in case)
- Missing openapi desc, doc
- SDK

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (TBonnin) [reply packages/server/lib/controllers/connect/postSessions.ts]: 🙈

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (TBonnin) [reply packages/server/lib/controllers/connect/postSessions.ts]: changed it but it is a bit ridiculous to have to repeat the same 10 lines on every single endpoint

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (TBonnin) [reply packages/server/lib/controllers/connect/postSessions.ts]: let's not call it profileId if it is not an id then. It might be confusing to users if we consider `profileId + orgId` to be the unique key

EDIT: I had misunderstood your comment. We are on the same page. We leave it to the nango users to decide what's an unique profile and we update profile attributes if they have changed for existing profile ids

File: packages/types/lib/linkedProfile/index.ts
Comment (TBonnin) [reply packages/types/lib/linkedProfile/index.ts]: noted

**Comment by TBonnin** [comment]
@bodinsamuel thank you for the review. I have addressed all your comments I think.

Regarding the missing openapi specs/doc and sdk changes, I wanted to first fully agree on the naming, input and output shapes, etc... before I start documenting things and adding to the sdk. I will do all of that in a subsequent PR

**Comment by bastienbeurier** [comment]
Kicking off the conversation about naming. 

@TBonnin proposed the term "linked profile": 
```
{
  "linkedProfile": {
     "profileId": "1jkhfz2FEuihojhfsdkjh29842", // Required
     "email": "a@a.com", // Required but for domain
     "displayName": "Mr A." // Optional
     "organization": { // Optional
        "organizationId": "foi", // Optional
        "displayName": "My org.", // Optional
     }
  }
}
```

I would propose using the terms "end user" and "end user organization", so the payload would look like: 
```
{
  "end_user_settings": {
     "id": "1jkhfz2FEuihojhfsdkjh29842",
     "email": "a@a.com",
     "display_name": "Mr A.",
     "org_id": "foi",
     "org_display_name": "My org"
  }
}
```
or 
```
{
  "settings": {
     "end_user_id": "1jkhfz2FEuihojhfsdkjh29842",
     "end_user_email": "a@a.com",
     "end_user_display_name": "Mr A.",
     "end_user_org_id": "foi",
     "end_user_org_display_name": "My org"
  }
}
```

@TBonnin @bodinsamuel @rguldener @khaliqgant what do you think?

Other naming detail: since we call the auth UI "Nango Connect", I would use the term "connect session" (and "connect session token" for the actual token) rather than "connection session", because this session only lives while establishing the connection, not for the whole lifespan of the connection.

**Comment by bodinsamuel** [comment]
> linkedProfile | end_user

end_user is confusing because it depends on your POV, if they are also B2B2C then end_user is a different one. That doesn't mean linkedProfile is way more explicit but at least it's POV free, and makes sense to me.

> flat vs deep object

I also prefer deep objects. Flat object is not easier to write, harder to read, annoying to document, and annoying to strictly type (TS type). In our codebase, it will be stored flat so DX over code here. 

> settings 

Too generic.


> Other naming detail: since we call the auth UI "Nango Connect", I would use the term "connect session" (and "connect session token" for the actual token) rather than "connection session", because this session only lives while establishing the connection, not for the whole lifespan of the connection.

Agreed indeed

**Comment by rguldener** [comment]
Agree with the others that `settings` is generic and doesn't fit well here. To me these are `properties` of an object (user or company).

Regarding the entity name: I would go for `user` and `organization`.
I briefly checked and its what [Posthog](https://posthog.com/docs/product-analytics/identify), [Amplitude](https://amplitude.com/docs/apis/analytics/identify), [Intercom](https://developers.intercom.com/installing-intercom/web/attributes-objects) and Mixpanel also do. I think it is likely our users are familiar with these and if it works for them I think it can work for us.

**Comment by bastienbeurier** [comment]
`user` and `organization` works for me. 

No conviction around `flat` vs. `deep` objects. 

@TBonnin the params will be in snake case, right? 

What about: 
```
{
  "user_properties": {
     "id": "",
     "email": "",
     "display_name": ""
  },
  "organization_properties": {
    "id": "",
    "display_name": ""
  }
}
```

or 

```
{
  "user_properties": {
     "id": "",
     "email": "",
     "display_name": "",
     "organization": {
       "id": "",
       "display_name": ""
    }
  },
  
}
```

File: packages/server/lib/controllers/connect/postSessions.ts
Comment (bodinsamuel) [reply packages/server/lib/controllers/connect/postSessions.ts]: yes, very happy to iterate on something similar to validateRequest, just need to follow what we previously set to avoid breaking change. It's mostly a matter of copying the code and changing the response

File: packages/shared/flows.yaml
Comment (bodinsamuel) [inline packages/shared/flows.yaml]: Merge is incorrect

**Review by bodinsamuel — APPROVED** [review]
Still waiting for final decision on naming but looks good 💪🏻

**Comment by rguldener** [comment]
The first one, with user & org on the same level, implies to me that they are independent.

The second, with org in the user, implies to me that the org is linked to the user with a 1:n

On deep vs not: the analytics tools all seem to have pretty flat structures. Not saying we have to copy, but might be worth a quick look at what they do and why.

**Comment by TBonnin** [comment]
the name `user` seems to be the most popular. I'd then prefer to go with @bastienbeurier proposal: `end_user`. 
_Why not simply `user`?_ Because there is also a notion of nango `user` and it might become very confusing very fast, especially for self-hosted customers.

so it will looks like this, without nesting organization to make it more extensible because the data model is simpler now but I wouldn't be surprised if it becomes more complex over time.
```
{
  "end_user": { // required
     "id": "",
     "email": "",
     "display_name": "", 
  }, 
  "organization": { // optional
     "id": "",
     "display_name": ""
  }
}
```

File: packages/types/lib/connect/api.ts
Comment (bodinsamuel) [inline packages/types/lib/connect/api.ts:21]: Old stuff

**Review by bodinsamuel — APPROVED** [review]
💯

File: packages/types/lib/connect/api.ts
Comment (TBonnin) [reply packages/types/lib/connect/api.ts:21]: yes. It is fixed in the other PR https://github.com/NangoHQ/nango/pull/2776/files#diff-bb1a9436ed436eaa04424f761fb7e7983183af72d751f15bb824395481e4adacL21

**Pull request merged**
